Index: src/game/SpellAuras.cpp
===================================================================
--- src/game/SpellAuras.cpp    (revision 976)
+++ src/game/SpellAuras.cpp    (working copy)
@@ -1700,6 +1700,12 @@
    if(apply)
    {
        SetPositive();
+        if (GetSpellProto()->Id == 28880)
+        {
+            sEventMgr.AddEvent(this, &Aura::EventPeriodicHeal,(uint32)(mod->m_amount+3*(GetUnitCaster()->getLevel()-1)),
+            EVENT_AURA_PERIODIC_HEAL,    GetSpellProto()->EffectAmplitude[mod->i],0);
+        }
+        else
        sEventMgr.AddEvent(this, &Aura::EventPeriodicHeal,(uint32)mod->m_amount,
            EVENT_AURA_PERIODIC_HEAL,    GetSpellProto()->EffectAmplitude[mod->i],0);
    }