Index: src/game/SpellAuras.cpp
===================================================================
--- src/game/SpellAuras.cpp    (revision 981)
+++ src/game/SpellAuras.cpp    (working copy)
@@ -1719,6 +1719,8 @@
            bonus += float2int32(((Player*)c)->SpellHealDoneByInt[m_spellProto->School] * ((Player*)c)->GetUInt32Value(UNIT_FIELD_STAT3));
            bonus += float2int32(((Player*)c)->SpellHealDoneBySpr[m_spellProto->School] * ((Player*)c)->GetUInt32Value(UNIT_FIELD_STAT4));
        }
+    bonus += c->HealDoneMod[GetSpellProto()->School];
+    bonus += m_target->HealTakenMod[GetSpellProto()->School];

    if(bonus)
    {
@@ -1732,11 +1734,9 @@
            bonus=bonus/ticks;
        }else bonus = 0;
    }
+    int add = bonus+amount;
+    add += float2int32(add*(m_target->HealTakenPctMod[GetSpellProto()->School]+c->HealDonePctMod[GetSpellProto()->School]/100.0f));
    
-    bonus+=m_target->HealTakenMod[GetSpellProto()->School];
-        
-    bonus += float2int32(m_target->HealTakenPctMod[GetSpellProto()->School]*amount);
-    int add = amount+bonus;
    uint32 newHealth = m_target->GetUInt32Value(UNIT_FIELD_HEALTH) + (uint32)add;
    
    if(newHealth <= m_target->GetUInt32Value(UNIT_FIELD_MAXHEALTH))