Index: SpellAuras.cpp
===================================================================
--- SpellAuras.cpp    (revision 1026)
+++ SpellAuras.cpp    (working copy)
@@ -5881,6 +5881,8 @@
        break;

    case SMT_DUMMY:
+        if(GetSpellProto()->dummy==457) AffectedGroups=1;
+        if(GetSpellProto()->dummy==86) AffectedGroups=256;
        SendModifierLog(&m_target->SM_FDummy,val,AffectedGroups,mod->m_miscValue);
        break;

Index: SpellEffects.cpp
===================================================================
--- SpellEffects.cpp    (revision 1026)
+++ SpellEffects.cpp    (working copy)
@@ -1672,6 +1672,11 @@
        if(motherspell)
            totalEnergy = (motherspell->EffectBasePoints[0]+1)*ProcedOnSpell->manaCost/100;
    }
+    else if (m_spellInfo->Id==2687){
+        totalEnergy = curEnergy+damage;
+        if (p_caster->HasSpell(12818)) totalEnergy += 60;
+        if (p_caster->HasSpell(12301)) totalEnergy += 30;
+    }
    else  totalEnergy = curEnergy+damage;
    if(totalEnergy > maxEnergy)
        unitTarget->SetUInt32Value(POWER_TYPE,maxEnergy);