Index: src/game/Spell.cpp
===================================================================
--- src/game/Spell.cpp    (revision 982)
+++ src/game/Spell.cpp    (working copy)
@@ -184,7 +184,7 @@
    if(u_caster && u_caster->GetCurrentSpell() == this)
        u_caster->SetCurrentSpell(NULL);
    
-    if(cancastresult == -1)
+    if(cancastresult == -1 && !failed)
        RemoveItems();
}