Index: src/game/Spell.cpp
===================================================================
--- src/game/Spell.cpp    (revision 1052)
+++ src/game/Spell.cpp    (working copy)
@@ -2074,11 +2074,16 @@
        case SPELL_RANGED_THROW:  // throw
            {
                   Item * it=p_caster->GetItemInterface()->GetInventoryItem(EQUIPMENT_SLOT_RANGED);
-                if(it)
-                {
-                    ip = it->GetProto();
-                       p_caster->GetItemInterface()->RemoveItemAmt(it->GetEntry(),1);
-                }
+if(it)
+{
+ip = it->GetProto();
+/* Throwing Weapon Patch by Supalosa
+p_caster->GetItemInterface()->RemoveItemAmt(it->GetEntry(),1);
+(Supalosa: Instead of removing one from the stack, remove one from durability)
+We don't need to check if the durability is 0, because you can't cast the Throw spell if the thrown weapon is broken, because it returns "Requires Throwing Weapon" or something.
+*/
+it->SetDurability( it->GetDurability() - 1 );
+}
            }break;
        default:
            {