Index: src/game/SpellEffects.cpp
===================================================================
--- src/game/SpellEffects.cpp	(revision 1014)
+++ src/game/SpellEffects.cpp	(working copy)
@@ -235,8 +235,9 @@
 	case 19443:
 		{
 		}break;
-	/*case 18788: //Demonic Sacrifice (508745)
-		switch(pPet->GetEntry())
+	case 18788: //Demonic Sacrifice (508745)
+		uint32 spellid1 = 0;
+		switch(unitTarget->GetEntry())
 		{
 			case 416: //Imp
 			{   
@@ -256,10 +257,20 @@
 			{
 				spellid1 = 18791;
 			}break;
-		}*/
+			case 17252: //felguard
+			{
+				spellid1 = 35701;
+			}break;
+		}
+		//now caster gains this buff
+		if (spellid1 && spellid1 != 0)
+		{
+			u_caster->CastSpell(u_caster, sSpellStore.LookupEntry(spellid1), true);
+		}
 	}
 
-	m_caster->SpellNonMeleeDamageLog(unitTarget, m_spellInfo->Id, unitTarget->GetUInt32Value(UNIT_FIELD_HEALTH), true);
+	//instant kill effects don't have a log
+	//m_caster->SpellNonMeleeDamageLog(unitTarget, m_spellInfo->Id, unitTarget->GetUInt32Value(UNIT_FIELD_HEALTH), true);
 }
 
 void Spell::SpellEffectSchoolDMG(uint32 i) // dmg school
@@ -2564,6 +2575,12 @@
 	CreatureInfo *ci = CreatureNameStorage.LookupEntry(m_spellInfo->EffectMiscValue[i]);
 	if(ci)
 	{
+		//if demonic sacrifice auras are still active, remove them
+		p_caster->RemoveAura(18789);
+		p_caster->RemoveAura(18790);
+		p_caster->RemoveAura(18791);
+		p_caster->RemoveAura(18792);
+		p_caster->RemoveAura(35701);
 		Pet *summon = objmgr.CreatePet();
 		summon->SetInstanceID(m_caster->GetInstanceID());
 		summon->CreateAsSummon(m_spellInfo->EffectMiscValue[i], ci, NULL, u_caster, m_spellInfo, 1, 0);
@@ -3742,17 +3759,18 @@
 
 void Spell::SpellEffectSummonDemon(uint32 i)
 {
-	if(!p_caster || p_caster->getClass() != WARLOCK)
+	if(!p_caster/* ||  p_caster->getClass() != WARLOCK */) //summoning a demon shouldn't be warlock only, see spells 25005, 24934, 24810 etc etc
 		return;
 	Pet *pPet = p_caster->GetSummon();
 	if(pPet)
 	{
 		pPet->Dismiss(false);
 	}
-   
+
 	CreatureInfo *ci = CreatureNameStorage.LookupEntry(m_spellInfo->EffectMiscValue[i]);
 	if(ci)
 	{
+
 		pPet = objmgr.CreatePet();
 		pPet->SetInstanceID(p_caster->GetInstanceID());
 		pPet->CreateAsSummon(m_spellInfo->EffectMiscValue[i], ci, NULL, p_caster, m_spellInfo, 1, 300000);
